src/FsAutoComplete.Core/NestedLanguages.fs

```diff
@@ -0,0 +1,318 @@
+
+         |]
+    // for the last part we have a single-character interpolation bracket on the left and the 'trailing' string elements on the right
+    | SynInterpolatedStringPart.String(range = range) when index = parts.Length - 1 ->
+      [|
+         // leading token adjustment - only a closing bracket }
+         range.WithEnd(range.Start.WithColumn(range.StartColumn + 1))
+
+         // trailing tokens adjustment
+         // GAP: we don't know how many """ to adjust for triple-quote interpolated string endings
+         match stringKind with
+         | SynStringKind.Regular ->
+           // 'regular' means trailing identifier "
+           range.WithStart(range.End.WithColumn(range.EndColumn - 1))
+         | SynStringKind.TripleQuote ->
+           // 'triple quote' means trailing identifier """
+           range.WithStart(range.End.WithColumn(range.EndColumn - 3))
+         // no such thing as verbatim interpolated strings
+         | SynStringKind.Verbatim -> () |]
+    // for all other parts we have a single-character interpolation bracket on the left and a trailing interpolation piece (which can include a format specifier) on the right
+    | SynInterpolatedStringPart.String(range = range) ->
+      [|
+         // leading token adjustment - only a closing bracket }
+         range.WithEnd(range.Start.WithColumn(range.StartColumn + 1))
+         // trailing token adjustment- only an opening bracket {
+         // GAP: this is the feature gap - we don't know about format specifiers here
+         range.WithStart(range.End.WithColumn(range.EndColumn - 1)) |])
+
+let private (|Ident|_|) (e: SynExpr) =
+  match e with
+  | SynExpr.Ident(ident) -> Some([ ident ])
+  | SynExpr.LongIdent(longDotId = SynLongIdent(id = ident)) -> Some ident
+  | _ -> None
+
+/// in order for nested documents to be recognized as their document types, the string quotes (and other tokens) need to be removed
+/// from the actual string content.
+let private removeStringTokensFromStringRange (kind: SynStringKind) (range: Range) : Range array =
+  match kind with
+  | SynStringKind.Regular ->
+    // we need to trim the double-quote off of the start and end
+    [| Range.mkRange range.FileName range.Start (range.Start.WithColumn(range.StartColumn + 1))
+       Range.mkRange range.FileName (range.End.WithColumn(range.EndColumn - 1)) range.End |]
+  | SynStringKind.Verbatim ->
+    // we need to trim the @+double-quote off of the start and double-quote off the end
+    [| Range.mkRange range.FileName range.Start (range.Start.WithColumn(range.StartColumn + 2))
+       Range.mkRange range.FileName (range.End.WithColumn(range.EndColumn - 1)) range.End |]
+  | SynStringKind.TripleQuote ->
+    // we need to trim the @+double-quote off of the start and double-quote off the end
+    [| Range.mkRange range.FileName range.Start (range.Start.WithColumn(range.StartColumn + 2))
+       Range.mkRange range.FileName (range.End.WithColumn(range.EndColumn - 1)) range.End |]
```

Comment:
@nojaf @vzarytovskii look at all of the nonsense string math I have to do in these two adjustment functions to get the 'real' content of the string :( Do you think the AST could/should learn any trivia here to make this kind of job easier?  There are a few different pieces that are troublesome:

* string/interpolated string start - this can be `"`, or `@"`, or `"""`, or now with interpolated strings any number of `$` followed by `"""`. this is untenable to get right currently.
* string/interpolated string end - this can be `"`, `"""`, or arbitrary `"` now. similarly bad :(
* format specifiers - these are relevant to formatting, but are not part of the logical string content. currently they are part of the reported string range and this means to do it 'right' I'd need to get the full string's content and do a subsequent parse
* string interpolation close/open brackets - these can be any number of `{`/`}` now with the expanded interpolation RFC

Essentially I'd love some kind of synthesized range for the _logical_ content of a string in the AST, because all of this math is very error prong.

Reply:
In Fantomas, we currently rely on the ranges of SynInterpolatedStringParts to restore the string. So, as long as these are accurate, we can grab what we need from ISourceText and no longer deal with this. In your case, it is indeed very messy to get right.